src/libsync/propagateremotedeleteencryptedrootfolder.cpp

```diff
@@ -0,0 +1,187 @@
+
+    qCDebug(PROPAGATE_REMOVE_ENCRYPTED_ROOTFOLDER) << "It's a root encrypted folder. Let's remove nested items first.";
+
+    metadata.removeAllEncryptedFiles();
+
+    qCDebug(PROPAGATE_REMOVE_ENCRYPTED_ROOTFOLDER) << "Metadata updated, sending to the server.";
+
+    auto job = new UpdateMetadataApiJob(_propagator->account(), _folderId, metadata.encryptedMetadata(), _folderToken);
+    connect(job, &UpdateMetadataApiJob::success, this, [this](const QByteArray& fileId) {
+        Q_UNUSED(fileId);
+        for (auto it = _nestedItems.constBegin(); it != _nestedItems.constEnd(); ++it) {
+            deleteNestedRemoteItem(it.key());
+        }
+    });
+    connect(job, &UpdateMetadataApiJob::error, this, &PropagateRemoteDeleteEncryptedRootFolder::taskFailed);
+    job->start();
+}
+
+void PropagateRemoteDeleteEncryptedRootFolder::slotDeleteNestedRemoteItemFinished()
+{
+    auto *deleteJob = qobject_cast<DeleteJob *>(QObject::sender());
+
+    ASSERT(deleteJob);
+
+    if (!deleteJob) {
+        return;
+    }
+
+    const QString encryptedFileName = deleteJob->property(encryptedFileNamePropertyKey).toString();
+
+    if (!encryptedFileName.isEmpty()) {
+        const auto nestedItem = _nestedItems.take(encryptedFileName);
+
+        if (nestedItem.isValid()) {
+            _propagator->_journal->deleteFileRecord(nestedItem._path, nestedItem._type == ItemTypeDirectory);
+            _propagator->_journal->commit("Remote Remove");
+        }
+    }
+
+    QNetworkReply::NetworkError err = deleteJob->reply()->error();
+
+    if (err != QNetworkReply::NoError) {
+        _errorString = deleteJob->errorString();
+    }
+
+    _item->_httpErrorCode = deleteJob->reply()->attribute(QNetworkRequest::HttpStatusCodeAttribute).toInt();
+    _item->_responseTimeStamp = deleteJob->responseTimestamp();
+    _item->_requestId = deleteJob->requestId();
+    _item->_errorString = _errorString;
+
+    if (err != QNetworkReply::NoError && err != QNetworkReply::ContentNotFoundError) {
+        _networkError = err;
+        taskFailed();
+        return;
+    }
+
+    // A 404 reply is also considered a success here: We want to make sure
+    // a file is gone from the server. It not being there in the first place
+    // is ok. This will happen for files that are in the DB but not on
+    // the server or the local file system.
+    if (_item->_httpErrorCode != 204 && _item->_httpErrorCode != 404) {
+        // Normally we expect "204 No Content"
+        // If it is not the case, it might be because of a proxy or gateway intercepting the request, so we must
+        // throw an error.
+        _errorString = tr("Wrong HTTP code returned by server. Expected 204, but received \"%1 %2\".")
+                .arg(_item->_httpErrorCode)
+                .arg(deleteJob->reply()->attribute(QNetworkRequest::HttpReasonPhraseAttribute).toString());
+        _item->_errorString = _errorString;
+        taskFailed();
```

Comment:
I wonder if that won't potentially lead to issues with the lock. Let's say _nestedItems.size() == 3. At that point that means that we got three delete job started. Now let's assume we got the first job failing for some reason.

If I'm not mistaken by calling taskFailed() this will lead to the remote delete job to end and be deleted, and so this object here being deleted after releasing the lock. So the two remaining delete jobs will potentially try to delete, supposedly it won't crash because their finished() signal will be disconnected but we'll have spurious delete attempts.

Maybe it's not too critical... it should converge and complete in the end, we'll just have spurious delete attempts in the logs. Another approach would be to store the error info, wait for all the deletes to complete and then claim the fail. Do we want that extra complexity?

Reply:
@er-vin I will change the code to make the **PropagateRemoteDeleteEncryptedRootFolder** wait until all it's nested items' DeleteJobs are finished, and then only fail if any of those jobs has failed and the result of a failure is stored within **PropagateRemoteDeleteEncryptedRootFolder** instance.